test/gtest/ucs/test_datatype.cc

```diff
@@ -1356,3 +1359,232 @@ UCS_MT_TEST_F(test_datatype_ptr_map_safe, safe_put, num_threads)
         }
     }
 }
+
+
+class test_piecewise_func : public test_datatype {
+protected:
+    struct segment {
+        size_t            start;
+        size_t            end;
+        ucs_linear_func_t func;
+    };
+
+    std::string get_segment_string(const segment &seg)
+    {
+        std::stringstream ss;
+        ss << "segment [" << seg.start << ", " << seg.end << "] : ";
+        ss << "func f(x) = " << seg.func.c << " + x*" << seg.func.m;
+
+        return ss.str();
+    }
+
+    void print_piecewise_func(const ucs_piecewise_func_t &func)
+    {
+        size_t start = 0;
+        ucs_piecewise_segment_t *seg;
+
+        UCS_TEST_MESSAGE << "piecewise func:";
+        ucs_list_for_each(seg, &func.head, list) {
+            segment test_segment = {start, seg->end, seg->func};
+            UCS_TEST_MESSAGE << "\t" << get_segment_string(test_segment);
+            start = seg->end + 1;
+        }
+    }
+
+    void compare_point_value(ucs_piecewise_func_t &piecewise_func,
+                             const ucs_linear_func_t &linear_func, size_t point)
+    {
+        ASSERT_EQ(ucs_piecewise_func_apply(&piecewise_func, point),
+                  ucs_linear_func_apply(linear_func, point))
+                  << "point is " << point;
+    }
+
+    size_t get_num_segments(const ucs_piecewise_func_t &func)
+    {
+        size_t num_ranges = 0;
+        ucs_piecewise_segment_t *seg;
+
+        ucs_list_for_each(seg, &func.head, list) {
+            ++num_ranges;
+        }
+        return num_ranges;
+    }
+
+    void check_signle_segment_func(const ucs_piecewise_func_t &func,
+                                   const ucs_linear_func_t &segment_func)
+    {
+        ASSERT_EQ(get_num_segments(func), 1);
+        ASSERT_TRUE(ucs_linear_func_is_equal(func.head_segment.func,
+                                             segment_func, 0));
+    }
+
+    void check_funcs_sum(ucs_piecewise_func_t &result,
+                         std::vector<ucs_piecewise_func_t> &funcs)
+    {
+        size_t seg_start = 0;
+        std::set<size_t> points;
+        ucs_piecewise_segment_t *seg;
+
+        for (auto &func : funcs) {
+            ucs_list_for_each(seg, &func.head, list) {
+                /* Test start, end and middle of the each segment */
+                points.insert(seg_start);
+                points.insert(seg->end);
+                points.insert(seg_start + (seg->end - seg_start) / 2);
+
```

Comment:
can `get_segment_points` be reused somehow?

Reply:
Done.